django/db/models/sql/query.py

```diff
@@ -1372,12 +1372,24 @@ def build_filter(
         if hasattr(filter_expr, "resolve_expression"):
             if not getattr(filter_expr, "conditional", False):
                 raise TypeError("Cannot filter against a non-conditional expression.")
+            pre_joins = self.alias_refcount.copy()
             condition = filter_expr.resolve_expression(
-                self, allow_joins=allow_joins, summarize=summarize
+                self, allow_joins=allow_joins, reuse=can_reuse, summarize=summarize
             )
             if not isinstance(condition, Lookup):
                 condition = self.build_lookup(["exact"], condition, True)
-            return WhereNode([condition], connector=AND), []
+            used_joins = {
+                k for k, v in self.alias_refcount.items() if v > pre_joins.get(k, 0)
+            }
+            require_outer = current_negated or (
+                isinstance(condition, IsNull)
+                and condition.rhs is True
+                and not current_negated
+            )
```

Comment:
I don't think we can get this logic any smarter until we add proper support for `Expression.nullable` which is the same reason why [ticket-25946](https://code.djangoproject.com/ticket/25946) is blocked for lookup transforms.

It's not a regression though since `LOUTER` was used systematically previously since `[]` was always returned.

Reply:
Per Anssi's [explanation on the topic](https://gist.github.com/akaariai/a4c1acbfedac0f2cbf3e)

> Currently, the only lookup that can match null values is isnull. However, it seems plausible that custom lookups can be used to create other lookups that match null values. This will be immediately obvious when expressions are allowed in the `filter()` clause. For example, `.filter(Equal(Coalesce('favourite_book__title', 'first_book__title'), 'Foo'))` must use left outer joins for the joins, as otherwise an author that has `favourite_book` with title
'Foo', but no first book would not be matched.

I think that's a strong enough rationale on why we can't merge the second commit here unfortunately.

In order to add proper support for this optimization we'd need to add a way to let the join promotion logic that an expression has special null-handling semantics (e.g. `IsNull`, `Coalesce`) and have it taint expressions that contain them like we do with `contains_aggregate` and similar attributes.

Reply:
Agreed, let's remove the 2nd commit.